src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,304 @@
+
+			if( m_preambles.Count != 0 ) {
+				throw new BugException( "left over preambles" );
+			}
+
+			if( m_methods.Count != 0 ) {
+				throw new BugException( "left over methods" );
+			}
+
+			return m_out.ToString();
+		}
+
+		private bool WriteChildren( SyntaxNode node ) {
+			// Early-out if we have no methods left to place.
+			if( m_methods.Count == 0 ) {
+				return false;
+			}
+
+			var preambleCountBefore = m_preambles.Count;
+
+			bool wroteSomething = false;
+
+			// First recurse into child types
+			foreach( var child in node.ChildNodes() ) {
+				bool wroteSomethingHere = child switch {
+					ClassDeclarationSyntax @class => WriteType( @class ),
+					StructDeclarationSyntax @struct => WriteType( @struct ),
+					InterfaceDeclarationSyntax @interface => WriteType( @interface ),
+
+					// Covers record + record struct:
+					RecordDeclarationSyntax @record => WriteType( @record, @record.ClassOrStructKeyword ),
+
+					// Covers file-scoped and block namespaces
+					BaseNamespaceDeclarationSyntax @namespace => WriteNamespace( @namespace ),
+
+					// Ignore any other child node
+					_ => false
+				};
+
+				if( wroteSomethingHere ) {
+					wroteSomething = true;
+				}
+
+				CheckPremableInvariant( wroteSomething, preambleCountBefore );
+			}
+
+			return wroteSomething;
+		}
+
+		private bool WriteNamespace( BaseNamespaceDeclarationSyntax @namespace ) {
+			SyntaxToken? closeBrace;
+
+			switch( @namespace ) {
+				case FileScopedNamespaceDeclarationSyntax ns:
+					m_preambles.AddLast(
+						new PieceOfSyntax[] {
+							ns.NamespaceKeyword,
+							ns.Name,
+							ns.SemicolonToken
+						}
+					);
+
+					closeBrace = null;
+					break;
+				case NamespaceDeclarationSyntax ns:
+					m_preambles.AddLast(
+						new PieceOfSyntax[] {
+							ns.NamespaceKeyword,
+							ns.Name,
+							ns.OpenBraceToken,
+							ns.Usings
+						}
+					);
+
+					closeBrace = ns.CloseBraceToken;
+					break;
+				default:
+					throw new BugException( "unexpected namespace style: " + @namespace.Kind() );
+			}
+
+			bool wroteSomething = WriteChildren( @namespace );
+
+			CloseBraceIfNecessary( wroteSomething, closeBrace );
+
+			return wroteSomething;
+		}
+
+		private bool WriteType( TypeDeclarationSyntax decl )
+			=> WriteType( decl, NoneToken );
+
+		private bool WriteType(
+			TypeDeclarationSyntax decl,
+			SyntaxToken secondaryKeyword
+		) {
+			var preambleCountBefore = m_preambles.Count;
+
+			// This should either get used inside a call to WriteMethods either
+			// somewhere down the call to WriteChildren here or the direct call
+			// to WriteMethods here, or it will be discarded inside
+			// CloseBraceIfNecessary.
+			AddPartialTypeToPreamble(
+				decl,
+				keyword1: decl.Keyword,
+				// for "record struct" etc.
+				keyword2: secondaryKeyword,
+				identifier: decl.Identifier,
+				typeParameterList: decl.TypeParameterList,
+				constraintClauses: decl.ConstraintClauses,
+				openBrace: decl.OpenBraceToken
+			);

```

Comment:
Not desirable to only do this if I or my children have methods?

Reply:
Sorry -- I don't follow

Reply:
Okay sorry, I missed that CloseBraceIfNeccesary is throwing this away if it was unused. I didn't read the method because the name seemed simple. Rename to `CloseTypeOrDiscardPreamble` or similar?

Reply:
CloseBlockOr..., rather, I suppose.

Reply:
Went the other direction with `FinishBlock` (vaguer -- it clearly "does something" and you should read it to understand, etc.)